### PR TITLE
[core] Update `no-response` workflow

### PR DESCRIPTION
**Pull request by MBilalShafi**
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

A follow-up to the https://github.com/mui/mui-x/pull/10102 as per discussion in https://www.notion.so/mui-org/Improve-support-automation-workflow-c491041115fa4a8e97ef483c8df5bf97?pvs=4
(Currently following approach A as mentioned in the document)

New GitHub action tested on this test repo: https://github.com/MBilalShafi/test-github-action/issues/2

TODOs after PR merge:
- [ ] Rename label `status: waiting for follow-up` -> `status: waiting for author`
- [ ] Rename label `status: needs triage` -> `status: waiting for maintainer`

### CODE REVIEW

**Review by michelengelen — APPROVED** [review]

**Review by DanailH — CHANGES_REQUESTED** [review]
I think in order to merge this PR we need more input and agreement on the shaping page first. We don't want to take a decision to rename labels just to rename them again later because there was a better option that more people agreed to.

**Comment by MBilalShafi** [comment]
> I think in order to merge this PR we need more input and agreement on the shaping page first. We don't want to take a decision to rename labels just to rename them again later because there was a better option that more people agreed to.

Agreed, the purpose of the PR is to let the readers visualize the change more in terms of "code". I'd prefer to have agreement and approvals from more people before the merge.

File: .github/workflows/no-response.yml
Comment (oliviertassinari) [inline .github/workflows/no-response.yml:27]: Shorter?

```suggestion
          responseRequiredLabel: "status: waiting for author"
```

```suggestion
          responseRequiredLabel: "status: need author's response"
```

File: .github/workflows/no-response.yml
Comment (oliviertassinari) [inline .github/workflows/no-response.yml:29]: Shorter?

```suggestion
          optionalFollowupLabel: "status: waiting for maintainer"
```

```suggestion
          optionalFollowupLabel: "status: need maintainer's response"
```

**Review by oliviertassinari — COMMENTED** [review]
This iteration makes me think about how we handle ticket statuses in Zendesk. In Zendeks, there are a couple of issue statutes and automation that come with them: https://www.notion.so/mui-org/Zendesk-de8ec63e1209475e8540dd5bb3097943?pvs=4#872f7db201d54126880b757d03033f91. I think it's helpful to see this as a state machine with state transitions.

Screenshot 2023-09-27 at 18 59 02

Today, the Zendesk status would be mapped like this to GitHub:

1. "New" is `status: need triage`
2. "Open" is `status: need triage`
3. "Pending" is `status: waiting for follow-up`, in Zendesk after 10 days tickets get moved to Solved automatically. If the person comments, they are moved back to "Open"
4. "On-hold" is the natural state of things, no labels
5. "Solved" is issue **closed**
6. "Closed" is issue **closed** or **closed** & **locked** or as **closed** & **unplanned**. But we don't lock in GitHub, in Zendesk it gets Closed after 7 days.

Something feels fundamentally strange with `status: waiting for maintainer's response`. It sounds to me that it's what 4. is for, so we shouldn't need a label. However, when we look at the use cases and why we are introducing it, it sounds like it's actually not 4 we try to communicate but 2. What if we had something more specific, like `status: need triage follow-up`. Or even nothing simply `status: need triage`? What problem does a different label solve?

**Edit**:

Oh actually, I'm not accurate, we are also using `status: waiting for follow-up` in 3. when we want to move to 5, and not just 6. We use this to get confirmation before closing the ticket as solved. So someone commenting to reopen wouldn't mean we should go back to 2. but to 4. 

I suspect we would want to disambiguate these two cases with specific labels:

- Have `status: needs more information` for when we move to state to 3 and when we want to go back to state 2.
- Have `status: needs confirmation` for when we move the state to 5. but without closing the issue yet, only going to 6. after a few days.

File: .github/workflows/no-response.yml
Comment (romgrk) [reply .github/workflows/no-response.yml:29]: My proposition in the notion doc is:
`status: waiting for maintainer`
`status: waiting for author`

Which is shorter but still clear.

**Comment by romgrk** [comment]
> Have `status: needs confirmation` for when we move from state 4. to 5. but without closing the issue yet, only going to 6.

I don't think we need that label. If it looks solved, we ask them to confirm and tag with `waiting for followup`, if they don't reply it gets auto-closed. We don't really need to differentiate issues in all those states.

I think we should keep github issues & labels as simple as possible, unless we have a clear need to make things more complex. I'm happy with just `waiting for author/maintainer`. We could even replace `triage` with `waiting for maintainer`, both mean the same: we need to look at them. Being able to find & track issues that need attention was the main concern when we discussed the label changes.

File: .github/workflows/no-response.yml
Comment (MBilalShafi) [reply .github/workflows/no-response.yml]: Looks good, implemented the change!

File: .github/workflows/no-response.yml
Comment (joserodolfofreitas) [inline .github/workflows/no-response.yml:21]: Should we use a "general" user for GitHub actions?

**Review by joserodolfofreitas — APPROVED** [review]
Great solution for the naming, and thank you for leading this effort, @MBilalShafi!

File: .github/workflows/no-response.yml
Comment (MBilalShafi) [reply .github/workflows/no-response.yml:21]: Not sure what you mean by `general` user. The action needed to be modified, so I forked the base one and published an updated one with a slightly different name.
Should we move it under https://github.com/mui ?

Comment (romgrk) [reply .github/workflows/no-response.yml:21]: I don't see a difference between it being on an external user's profile or a maintainer's profile. External user more dangerous than maintainer though (git repos are not immutable, so in theory it would be possible to craft a commit that has the same hash but different malicious code). But I'm fine with this as it is.

Alternatively I also like what we do in other scripts with `actions/github-script` and inline code in the yaml action definition, keeps things tidy: https://github.com/mui/mui-x/blob/2e50f761a0325b95d33d52faad0b6d2fccb21e8a/.github/workflows/issue-cleanup.yml#L16-L52

Comment (romgrk) [reply .github/workflows/no-response.yml:21]: No let's not waste time no that, it's not worth it. But for from-scratch actions it's probably better to have it inline.

**Comment by MBilalShafi** [comment]
@mui/explore team, it'll be great if someone from you have a look over this change and see if it's ok with you as it'll equally impact both the teams.
Thank you 🙂

**Review by romgrk — APPROVED** [review]

**Review by cherniavskii — APPROVED** [review]

**Review by LukasTy — APPROVED** [review]
Great initiative! 💯 
I'm all for experimenting on this regard, especially given the presented problems. 👌

File: .github/workflows/no-response.yml
Comment (oliviertassinari) [reply .github/workflows/no-response.yml]: > status: waiting for author

Could this be confused with "author" as the author of the code?

Would "waiting for user" be clearer? As I understand it, the GitHub Action nor maintainers truly need the author to answer, but more have confirmation from users.

File: .github/workflows/no-response.yml
Comment (oliviertassinari) [inline .github/workflows/no-response.yml:28]: I got confused by this comment. Would it be clearer like this?

```suggestion
          # The label that will be added back when the `responseRequiredLabel` is removed due to the author responding back.
```

**Comment by oliviertassinari** [comment]
The move makes sense 👍 , regarding next step:

@MBilalShafi @michelengelen

*Now*, Can we:

- Delete the needs triage level for MUI X: https://github.com/mui/mui-x/labels?q=need+triage
- Update the needs triage KPI in for MUI X https://www.notion.so/mui-org/KPIs-1ce9658b85ce4628a2a2ed2ae74ff69c?pvs=4#c0c162d3503649c89b3ea97d8eb86c62

*Once we validate the workflow in MUI X, in a few weeks*, Can we:

- Update the needs triage KPI in for the whole org https://www.notion.so/mui-org/KPIs-1ce9658b85ce4628a2a2ed2ae74ff69c?pvs=4#c0c162d3503649c89b3ea97d8eb86c62
- Fix the Toolpad apps Greg uses to do the first tier level supports that relies on this label https://www.notion.so/mui-org/GitHub-community-issues-PRs-Tier-1-12a84fdf50e44595afc55343dac00fca.
- Update https://github.com/mui/material-ui, https://github.com/mui/mui-toolpad, https://github.com/mui/mui-private, https://github.com/mui/mui-public, https://github.com/mui/mui-design-kits to reflect this new workflow?
- Update notion about the new name https://www.notion.so/mui-org/Technical-support-80c9c7e0af614af68c1abb7120f634c8?pvs=4#ade93334755f49a58c51c2d2399eaf05, https://www.notion.so/mui-org/GitHub-labels-af1ef18d3f534f1dbc086aac87b819e8?p=1d641537fe044358b455837cba97031f&pm=s, etc.
- Maybe? Document the support state machine that we have in place in https://www.notion.so/mui-org/Technical-support-80c9c7e0af614af68c1abb7120f634c8, like we have in https://www.notion.so/mui-org/Zendesk-de8ec63e1209475e8540dd5bb3097943

Thanks

**Comment by oliviertassinari** [comment]
**Off-topic**

A fun hack used by Metabase, they prefix some labels with a dot so it shows first https://github.com/metabase/metabase/issues

Screenshot 2023-10-18 at 17 44 05

**Comment by michelengelen** [comment]
> The move makes sense 👍 , regarding next step:
> 
> @MBilalShafi @michelengelen
> 
> _Now_, Can we:
> 
> * Update the needs triage KPI in for MUI X https://www.notion.so/mui-org/KPIs-1ce9658b85ce4628a2a2ed2ae74ff69c?pvs=4#c0c162d3503649c89b3ea97d8eb86c62
> 
> _Once we validate the workflow in MUI X, in a few weeks_, Can we:
> 
> * Delete the needs triage label for MUI X: https://github.com/mui/mui-x/labels?q=need+triage
> * Update the needs triage KPI in for the whole org https://www.notion.so/mui-org/KPIs-1ce9658b85ce4628a2a2ed2ae74ff69c?pvs=4#c0c162d3503649c89b3ea97d8eb86c62
> * Fix the Toolpad apps Greg uses to do the first tier level supports that relies on this label https://www.notion.so/mui-org/GitHub-community-issues-PRs-Tier-1-12a84fdf50e44595afc55343dac00fca.
> * Update https://github.com/mui/material-ui, https://github.com/mui/mui-toolpad, https://github.com/mui/mui-private, https://github.com/mui/mui-public, https://github.com/mui/mui-design-kits to reflect this new workflow for the whole organization
> * Update notion about the renamed "needs triage" label, in https://www.notion.so/mui-org/Technical-support-80c9c7e0af614af68c1abb7120f634c8?pvs=4#ade93334755f49a58c51c2d2399eaf05, https://www.notion.so/mui-org/GitHub-labels-af1ef18d3f534f1dbc086aac87b819e8?p=1d641537fe044358b455837cba97031f&pm=s, etc.
> * Maybe? Document the support state machine that we have in place in https://www.notion.so/mui-org/Technical-support-80c9c7e0af614af68c1abb7120f634c8, like we have in https://www.notion.so/mui-org/Zendesk-de8ec63e1209475e8540dd5bb3097943
> 
> Thanks

@MBilalShafi I'll take care of the updates next week, when I am back!

**Comment by oliviertassinari** [comment]
@michelengelen thanks for opening a new GitHub issue 👌, I should have done this.

**Comment by michelengelen** [comment]
> @michelengelen thanks for opening a new GitHub issue 👌, I should have done this.

All good! As I wanted to follow the next steps once we are comfortable it is not a big deal!
😂

**Pull request merged**
